src/main/java/UserRepository.java

```diff
@@ -6,5 +10,67 @@
  * uniqueness of usernames to prevent duplicates.
  */
 public class UserRepository {
-    // TODO implement class
+
+    private List<User> users;
+
+    public UserRepository() {
+        this.users = new ArrayList<>();
+    }
+
+    public void addUser(User user) {
+        if (isUnique(user)) {
+            users.add(user);
+        } else {
+            System.out.println("User is already in repository.");
+        }
+    }
+
+    public User getUserById(UUID id) {
+        for (User user : users) {
+            if (user.getId() == id) {
+                return user;
+            }
+        }
+        System.out.println("User does not exist");
+        return null;
+    }
+
+    public User getUserByUsername(String username) {
+        for (User user : users) {
+            if (user.getName().equals(username)) {
+                return user;
+            }
+        }
+        System.out.println("Username does not exist");
+        return null;
+    }
+
+    public void updateUser(User oldUser, User updatedUser) {
+        if (isUnique(updatedUser)) {
+            int index = users.indexOf(oldUser);
+            users.set(index, updatedUser);
+        } else {
+            System.out.println("Username already exists");
+        }
+    }
+
+    public void removeUserById(UUID id) {
+        User user = getUserById(id);
+        if (isUnique(user)) {
```

Comment:
You don't need this check, the user is for sure unique as you don't allow 2 people with the same username. 
I guess it might have sense in an async scenario with different threads, but in that case you wouldn't use a normal List, so yeah I'd remove this check.

Reply:
also, NPE here as user could be null